core/src/main/scala/com/salesforce/op/stages/impl/tuning/OpCrossValidation.scala

```diff
@@ -57,16 +57,23 @@ private[op] class OpCrossValidation[M <: Model[_], E <: Estimator[_]]
   private def findBestModel(
     folds: Seq[ValidatedModel[E]]
   ): ValidatedModel[E] = {
-    val metrics = folds.map(_.metrics).reduce(_ + _)
-    blas.dscal(metrics.length, 1.0 / numFolds, metrics, 1)
-    val ValidatedModel(est, _, _, grid) = folds.head
-    log.info(s"Average cross-validation for $est metrics: {}", metrics.toSeq.mkString(","))
-    val (bestMetric, bestIndex) =
-      if (evaluator.isLargerBetter) metrics.zipWithIndex.maxBy(_._1)
-      else metrics.zipWithIndex.minBy(_._1)
-    log.info(s"Best set of parameters:\n${grid(bestIndex)}")
+    require(folds.map(_.model.uid).toSet.size == 1) // Should be called only on instances of the same model
+    val gridCounts = folds.map(_.grids.map(_ -> 1).toMap).reduce(_ + _)
+    val maxFolds = gridCounts.maxBy(_._2)._2
```

Comment:
For better readability please replace `_._2` in `val maxFolds = gridCounts.maxBy(_._2)._2` with `val (_, maxFolds) = gridCounts.maxBy { case (_, folds) => folds }`
Same in a few operations below.